### PR TITLE
Fix SIGILL on nightly rustc

### LINKED ISSUE #99: #90 Seems to have come back again
Issue #90 seems to have come back again...

The issue is really, really weird though; when I build in debug mode, I don't have this bug, but when I rebuild (completely from scratch) in release mode, it comes back to bite me.  Here's the output from the `lldb`.  Compiler info, etc., are in the Meta section below.

```
$ lldb target/release/disc_model_simulator_driver -- ../scenarios.toml 
(lldb) target create "target/release/disc_model_simulator_driver"
Current executable set to 'target/release/disc_model_simulator_driver' (x86_64).
(lldb) settings set -- target.run-args  "../scenarios.toml"
(lldb) run
Process 32688 launched: '/home/cfkaran2/Documents/Dissertation/disc_model_simulator_driver/target/release/disc_model_simulator_driver' (x86_64)
ERROR: "/home/cfkaran2/Documents/Dissertation/disc_model_simulator/src/debug_memory_use.rs": 212: Debug_memory_logging_thread has started.
Note: Release mode will improve performance greatly.
    e.g. use `cargo run --release`
Process 32688 stopped
* thread #1, name = 'disc_model_simu', stop reason = signal SIGILL: illegal instruction operand
    frame #0: 0x0000555555947d5d disc_model_simulator_driver`x11_dl::xlib_xcb::Xlib_xcb::open::h23f529574679c79f at link.rs:0:7
   1   	// x11-rs: Rust bindings for X11 libraries
   2   	// The X11 libraries are available under the MIT license.
   3   	// These bindings are public domain.
   4   	
   5   	use std::ffi::{ CStr, CString };
   6   	use std::path::Path;
   7   	use std::os::raw::{ c_char, c_void };
(lldb) bt
error: need to add support for DW_TAG_base_type 'char' encoded with DW_ATE = 0x8, bit_size = 32
* thread #1, name = 'disc_model_simu', stop reason = signal SIGILL: illegal instruction operand
  * frame #0: 0x0000555555947d5d disc_model_simulator_driver`x11_dl::xlib_xcb::Xlib_xcb::open::h23f529574679c79f at link.rs:0:7
    frame #1: 0x0000555555865a0a disc_model_simulator_driver`winit::platform::platform::x11::xdisplay::XConnection::new::h6b68c0a556c991c9(error_handler=Option<unsafe extern "C" fn(*mut x11_dl::xlib::_XDisplay, *mut x11_dl::xlib::XErrorEvent) -> i32> @ r14) at xdisplay.rs:39:23
    frame #2: 0x00005555558759f8 disc_model_simulator_driver`spin::once::Once$LT$T$GT$::call_once::hc944d05a52b546ee [inlined] _$LT$winit..platform..platform..X11_BACKEND$u20$as$u20$core..ops..deref..Deref$GT$::deref::__static_ref_initialize::h33afb142c7d0d9df at mod.rs:54:19
    frame #3: 0x00005555558759e6 disc_model_simulator_driver`spin::once::Once$LT$T$GT$::call_once::hc944d05a52b546ee [inlined] core::ops::function::FnOnce::call_once::h03d3fd96eef4dec8 at function.rs:231
    frame #4: 0x00005555558759e6 disc_model_simulator_driver`spin::once::Once$LT$T$GT$::call_once::hc944d05a52b546ee(self=0x0000555555c311a0, builder=<unavailable>) at once.rs:110
    frame #5: 0x00005555558839e0 disc_model_simulator_driver`winit::platform::platform::EventsLoop::new_x11::hdb92276b71d7487a [inlined] lazy_static::lazy::Lazy$LT$T$GT$::get::ha7a0c63009b0f450 at core_lazy.rs:21:8
    frame #6: 0x00005555558839d4 disc_model_simulator_driver`winit::platform::platform::EventsLoop::new_x11::hdb92276b71d7487a [inlined] _$LT$winit..platform..platform..X11_BACKEND$u20$as$u20$core..ops..deref..Deref$GT$::deref::__stability::h5f73dcc3ea994ef9 at <::lazy_static::__lazy_static_internal macros>:12
    frame #7: 0x00005555558839d4 disc_model_simulator_driver`winit::platform::platform::EventsLoop::new_x11::hdb92276b71d7487a [inlined] _$LT$winit..platform..platform..X11_BACKEND$u20$as$u20$core..ops..deref..Deref$GT$::deref::h7fc2a0f514c058e5(self=<unavailable>) at <::lazy_static::__lazy_static_internal macros>:13
    frame #8: 0x00005555558839d4 disc_model_simulator_driver`winit::platform::platform::EventsLoop::new_x11::hdb92276b71d7487a at mod.rs:459
    frame #9: 0x000055555588361f disc_model_simulator_driver`winit::platform::platform::EventsLoop::new::he6b4ce1a408563a5 at mod.rs:440:28
    frame #10: 0x00005555558a003a disc_model_simulator_driver`winit::EventsLoop::new::h6f539370fc8eaf4c at lib.rs:250:25
    frame #11: 0x000055555571b9f4 disc_model_simulator_driver`ggez::context::ContextBuilder::build::hddca8b3bb3196779 at context.rs:69:26
    frame #12: 0x000055555571b89e disc_model_simulator_driver`ggez::context::ContextBuilder::build::hddca8b3bb3196779(self=ContextBuilder @ 0x00007fffffffc500) at context.rs:273
    frame #13: 0x000055555564df6d disc_model_simulator_driver`disc_model_simulator_driver::main::_$u7b$$u7b$closure$u7d$$u7d$::h3d4016fcf37dcd4f((null)=<unavailable>, scenario=Scenario @ 0x00007fffffffd8b0) at main.rs:291:37
    frame #14: 0x000055555564d442 disc_model_simulator_driver`disc_model_simulator_driver::main::h339ab0a862cd6123 [inlined] core::iter::traits::iterator::Iterator::for_each::_$u7b$$u7b$closure$u7d$$u7d$::h153c46f0df67cdea at iterator.rs:604:38
    frame #15: 0x000055555564d3ca disc_model_simulator_driver`disc_model_simulator_driver::main::h339ab0a862cd6123 [inlined] core::iter::traits::iterator::Iterator::fold::_$u7b$$u7b$closure$u7d$$u7d$::h739f9c6e8319c5f4 at iterator.rs:1685
    frame #16: 0x000055555564d3ca disc_model_simulator_driver`disc_model_simulator_driver::main::h339ab0a862cd6123 [inlined] core::iter::traits::iterator::Iterator::try_fold::h05f7e3384a01aa3c(self=<unavailable>) at iterator.rs:1573
    frame #17: 0x000055555564d356 disc_model_simulator_driver`disc_model_simulator_driver::main::h339ab0a862cd6123 [inlined] core::iter::traits::iterator::Iterator::fold::hd177379ce9516e2d(self=<unavailable>) at iterator.rs:1685
    frame #18: 0x000055555564d356 disc_model_simulator_driver`disc_model_simulator_driver::main::h339ab0a862cd6123 [inlined] core::iter::traits::iterator::Iterator::for_each::h10954f1698a2ed5f(self=IntoIter<disc_model_simulator_driver::scenario_parser::Scenario> @ 0x00000000027282b0) at iterator.rs:604
    frame #19: 0x000055555564d32c disc_model_simulator_driver`disc_model_simulator_driver::main::h339ab0a862cd6123 at main.rs:273
    frame #20: 0x00005555555dbe72 disc_model_simulator_driver`std::rt::lang_start::_$u7b$$u7b$closure$u7d$$u7d$::h4f4551107d3dd7b2 at rt.rs:64:33
    frame #21: 0x0000555555a0b9b3 disc_model_simulator_driver`std::panicking::try::do_call::hcfcd808f74618614 [inlined] std::rt::lang_start_internal::_$u7b$$u7b$closure$u7d$$u7d$::ha6cdc0674c227baf at rt.rs:49:12
    frame #22: 0x0000555555a0b9a7 disc_model_simulator_driver`std::panicking::try::do_call::hcfcd808f74618614 at panicking.rs:296
    frame #23: 0x0000555555a105da disc_model_simulator_driver`__rust_maybe_catch_panic at lib.rs:82:7
    frame #24: 0x0000555555a0c4bd disc_model_simulator_driver`std::rt::lang_start_internal::ha819bd148e81f2bd [inlined] std::panicking::try::h28299215a9cbdc2e at panicking.rs:275:12
    frame #25: 0x0000555555a0c47f disc_model_simulator_driver`std::rt::lang_start_internal::ha819bd148e81f2bd [inlined] std::panic::catch_unwind::h6d2adbec69dbd2ec at panic.rs:394
    frame #26: 0x0000555555a0c47f disc_model_simulator_driver`std::rt::lang_start_internal::ha819bd148e81f2bd at rt.rs:48
    frame #27: 0x0000555555651b08 disc_model_simulator_driver`main + 40
    frame #28: 0x00007ffff7c82b6b libc.so.6`__libc_start_main + 235
    frame #29: 0x00005555555a023a disc_model_simulator_driver`_start + 42
```

# Meta

```
Linux Dissertation 5.0.0-20-generic #21-Ubuntu SMP Mon Jun 24 09:32:09 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux

Distributor ID:	Ubuntu
Description:	Ubuntu 19.04
Release:	19.04
Codename:	disco

rustc 1.36.0 (a53f9df32 2019-07-03)
binary: rustc
commit-hash: a53f9df32fbb0b5f4382caaad8f1a46f36ea887c
commit-date: 2019-07-03
host: x86_64-unknown-linux-gnu
release: 1.36.0
LLVM version: 8.0

rustc 1.37.0-beta.2 (74e5a0d47 2019-07-08)
binary: rustc
commit-hash: 74e5a0d47ef4614524675f917e42b3fc45e8f759
commit-date: 2019-07-08
host: x86_64-unknown-linux-gnu
release: 1.37.0-beta.2
LLVM version: 8.0

rustc 1.38.0-nightly (0b680cfce 2019-07-09)
binary: rustc
commit-hash: 0b680cfce544ff9a59d720020e397c4bf3346650
commit-date: 2019-07-09
host: x86_64-unknown-linux-gnu
release: 1.38.0-nightly
LLVM version: 8.0

cargo 1.36.0 (c4fcfb725 2019-05-15)
release: 1.36.0
commit-hash: c4fcfb725b4be00c72eb9cf30c7d8b095577c280
commit-date: 2019-05-15

cargo 1.37.0-beta (4c1fa54d1 2019-06-24)
release: 1.37.0
commit-hash: 4c1fa54d10f58d69ac9ff55be68e1b1c25ecb816
commit-date: 2019-06-24

cargo 1.37.0-nightly (4c1fa54d1 2019-06-24)
release: 1.37.0
commit-hash: 4c1fa54d10f58d69ac9ff55be68e1b1c25ecb816
commit-date: 2019-06-24
```

### PR DESCRIPTION
**Pull request by ishitatsuyuki**
Fix #99, I only fixed the very issue interrupting my workflow, and I'll leave other unsoundness as-is.

### CODE REVIEW

File: x11-dl/src/link.rs [hunk 1]
```diff
@@ -55,12 +55,11 @@ macro_rules! x11_link {
         unsafe {
           let libdir = $crate::link::config::libdir::$pkg_name;
           let lib = try!($crate::link::DynamicLibrary::open_multi(libdir, &[$($lib_name),*]));
-          let mut this: ::std::mem::ManuallyDrop<$struct_name>
-              = ::std::mem::uninitialized();
+          let mut this = ::std::mem::MaybeUninit::<$struct_name>::uninit();
           let this_ptr = &mut this as *mut _ as *mut $struct_name;
```
Comment (bschwind) [inline x11-dl/src/link.rs]: Now that we're using `MaybeUninit`, it might be better to use `this.as_mut_ptr()` here

**Comment by bschwind** [comment]
You beat me to it! I made essentially the same change which fixed the particular issue I was hitting in #99, with the difference being that I used `.as_mut_ptr()`

I'm not sure if `&mut this as *mut _ as *mut $struct_name;` counts as "taking a reference to uninitialized memory", but if it does I would assume that also implies undefined behavior.

**Comment by ishitatsuyuki** [comment]
Yeah, that was a good catch! I adopted that suggestion.

(The code you mention only takes reference of `MaybeUninit` which is completely safe, and then it's turned into a raw pointer which the compiler won't mess with.)

**Comment by bschwind** [comment]
Looks good to me, and fixes the SIGILL I was encountering earlier.

**Comment by bjorn3** [comment]
> I'm not sure if &mut this as *mut _ as *mut $struct_name; counts as "taking a reference to uninitialized memory"

No, this is of type `MaybeUninit<Self>` which is explicitely allowed to be uninitialized (thats the purpose of it), taking a reference to it is thus allowed.

**Comment by RalfJung** [comment]
> I'm not sure if &mut this as *mut _ as *mut $struct_name; counts as "taking a reference to uninitialized memory", but if it does I would assume that also implies undefined behavior.

This is very subtle, and I first thought it does, but indeed the type of `&mut this` is `&mut MaybeUnint` and that reference can point to anything -- "initialized" is a type-dependent notion; an "initialized" `bool` must be 0x00 or 0x01 but OTOH a `MaybeUninit` is *always* "initialized" even if the memory is not.

Maybe we should pick different terms, not sure...

But either way, raw pointer casts should always be avoided, so `as_mut_ptr` is definitely preferable here.

**Comment by RalfJung** [comment]
Note that `MaybeUninit` is very recent. I definitely recommend to use it, but it means you need at least Rust 1.36.

As an interim solution, it might help to (a) use `mem::zeroed`, that's always preferable over `mem::uninitialized` (and if you are seeing the opposite effect, you should debug your code to find out where you are zero-initializing something like a reference or a function pointer that *may never be NULL*), and (b) fix this to use `Option`:

https://github.com/erlepereira/x11-rs/blob/7fe454564799c22e7bc1069758bd99f372aaf095/x11-dl/src/link.rs#L30-L31

**Comment by est31** [comment]
> it means you need at least Rust 1.36.

If this is a problem, I recommend adopting my maybe-uninit crate: It falls back to `std::mem::uninitialized` on older rustcs while just reexporting `MaybeUninit` on newer ones: https://crates.io/crates/maybe-uninit

**Comment by ishitatsuyuki** [comment]
Here you go. Maintainers: remove the latest commit if you don't like the extra dependency.

**Comment by RalfJung** [comment]
However, the `maybe-uninit` crate does not really avoid the UB in pre-Rust-1.36. (I think it uses `ManuallyDrop` there?) So I'd still recommend also turning the function pointers into `Option` to avoid making false promises to the optimizer.

**Comment by est31** [comment]
@RalfJung yes, you are right if fn pointers always have to be valid then `maybe-uninit::MaybeUninit` is using UB on older Rust versions. However, on older Rust versions, which won't change any more, this seems to not have been much of an issue, at least not in this particular instance. It only started to be an issue once Rust started changing stuff about `mem::uninitialized`.

In other words: you look at it in a non-constructive way and say "on older rustcs this crate is just as bad as not using it!" while I look at it in a constructive way and say "on newer rustcs this crate is doing the right thing while it doesn't regress older rustcs support".

In another thread you said that it took you two years to figure out this is a problem and two more years to provide a fix for it. And now the ecosystem should switch within six weeks? IMO the order of magnitude is wrong here.

That being said, I agree that `mem::zeroed` + `Option` is better than using `maybe-uninit`, at least in this instance.

**Comment by RalfJung** [comment]
I would say I look at it in a foundational/theoretical way and you look at it in a pragmatic way. ;) I think it is unfair to call my approach "non-constructive".

Note that using `ManuallyDrop` to hold invalid values *is* known to cause problems also in old compilers; IIRC `SmallVec` ran into that. So these concerns are not just theoretical.

> In another thread you said that it took you two years to figure out this is a problem and two more years to provide a fix for it. And now the ecosystem should switch within six weeks? IMO the order of magnitude is wrong here.

I (constructively) proposed a way to fix the problem here without having to depend on Rust 1.36. I am not sure what your problem with the proposal I made is -- or if your problem is not with the proposal, then I do not know what you are even criticizing here.

Sure, without `Option` it might still work, but we seem to agree that it is more correct with `Option`. (And *that* has been clear for more than two years, so there was ample time.  My interpretation is that the maintainer just was not aware that `fn()` types are non-NULL. This is understandable -- that invariant is not talked about very often, as `fn()` types are not widely used in Rust, and the two [lists](https://doc.rust-lang.org/nomicon/what-unsafe-does.html) of [UB](https://doc.rust-lang.org/reference/behavior-considered-undefined.html) that we have both forgot to list it as well.)

**Comment by est31** [comment]
I don't have a problem with the `Option` proposal. I guess it requires some unwraps here and there. It's fine.

**Comment by est31** [comment]
As for unconstructive vs theoretical, I just think that dismissing the entire `maybe-uninit` crate because it doesn't shine well in the theoretical light (which admittedly is true) is unfair because it does have pragmatic benefits. Overall there is a way out that we both agree is valid.

**Comment by RalfJung** [comment]
I didn't mean to dismiss it -- sorry if my comment came across as such. I was just pointing out that using the `Option` thing can still help if x11-rs decides to use that crate.

**Comment by est31** [comment]
@Daggerbot @erlepereira want to review this?

**Comment by erlepereira** [comment]
hey thanks for the ping and the comments/discussion everyone. 
Been meaning to jump onto this code. 

Just a short note, will be looking into this over this weekend.

**Comment by est31** [comment]
@erlepereira weekend's here!

**Comment by erlepereira** [comment]
The mem::zeroed + Option<T> does look the ideal fix here. On the MaybeUninit front,  a bit reluctant to introduce it just yet (min version needed) and reluctant to introduce a new dependency, it would be effectively forcing it on everyone else who use this. Though of course, if we did at this point it would likely be an interim measure anyway.

 @Daggerbot any preferences here? -- you know a lot more of the history around this code here

I cannot see any errors from switching to mem::zeroed on my end, still working on a proper solution using Option<T> (some test code breaks...). Continuing to look deeper  for a  workable solution  on that front.

_I do realize there is a ticking timeline and this crate has not been updated in a while. Hope to have a  fix soon, else this patch + dependency might have to be the way forward for now._

**Comment by ishitatsuyuki** [comment]
I think with `MaybeUninit` the UB problem is solved, modulo rustc version requirement - @Daggerbot does this address your concern?

**Comment by est31** [comment]
@Daggerbot @erlepereira is the current solution any good? The new code is only a "hack" on versions of Rust where there are no problems and on any further versions it uses the recommended `MaybeUninit`. I'd love to have a fix for this because it makes my app crash at startup :/. The additional dependency is very small, quick to compile and doesn't have any dependencies of its own. And it's only temporary until you choose to drop support for versions older than 1.36.0.

File: x11-dl/src/link.rs [hunk 2]
```diff
@@ -55,12 +55,11 @@ macro_rules! x11_link {
...
           let lib = try!($crate::link::DynamicLibrary::open_multi(libdir, &[$($lib_name),*]));
-          let mut this: ::std::mem::ManuallyDrop<$struct_name>
-              = ::std::mem::uninitialized();
-          let this_ptr = &mut this as *mut _ as *mut $struct_name;
+          let mut this = ::maybe_uninit::MaybeUninit::<$struct_name>::uninit();
+          let this_ptr = this.as_mut_ptr();
           ::std::ptr::write(&mut (*this_ptr).lib, lib);
           try!(Self::init(this_ptr));
-          Ok(::std::mem::ManuallyDrop::into_inner(this))
+          Ok(this.assume_init())
```
Comment (RalfJung) [inline x11-dl/src/link.rs]: So, it is guaranteed then that `Self::init` *will* put non-NULL values into all these function pointers?

**Comment by erlepereira** [comment]
As stated earlier, this code-base needs some attention. Yeah, for the interim I guess this works fine.

For the longer term, need to update this code, bringing up the supported version while incorporating the various ideas raised here and in other issues. Will bring all these ideas together in another thread / issues,  and find a workable way forward for this/other issues that maybe related.
 
Leaving this issue open until I merge the patch, release an updated crate with this patch.
 Eta later tonight/tomorrow.

**Pull request merged**